engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/VectorFromFunctionNode.java

```diff
@@ -0,0 +1,150 @@
+   * @param onProblems Can be either an atom of type {@code Problem_Behavior} or {@code No_Wrap}
+   *     type.
+   * @return Vector constructed from the given function.
+   */
+  abstract Object execute(
+      VirtualFrame frame, State state, long length, Function func, Object onProblems);
+
+  @Specialization
+  Object doIt(
+      VirtualFrame frame,
+      State state,
+      long length,
+      Function func,
+      Object onProblemsAtom,
+      @Cached("buildWithArity(1)") InvokeFunctionNode invokeFunctionNode,
+      @Cached("build()") AppendWarningNode appendWarningNode,
+      @CachedLibrary(limit = "3") WarningsLibrary warnsLib,
+      @CachedLibrary(limit = "3") TypesLibrary typesLib,
+      @Cached BranchProfile errorEncounteredProfile) {
+    var ctx = EnsoContext.get(this);
+    var onProblems = processOnProblemsArg(onProblemsAtom, typesLib);
+    var len = Math.toIntExact(length);
+    var nothing = ctx.getNothing();
+    var target = ArrayBuilder.newBuilder(len);
+    var errorsEncountered = 0;
+    for (int i = 0; i < len; i++) {
+      var value = invokeFunctionNode.execute(func, frame, state, new Long[] {(long) i});
+      Object valueToAdd = value;
+      if (value instanceof DataflowError err) {
+        errorEncounteredProfile.enter();
+        switch (onProblems) {
+          case IGNORE -> valueToAdd = nothing;
+          case REPORT_ERROR -> {
+            var mapErr = ctx.getBuiltins().error().makeMapError(i, err.getPayload());
+            return DataflowError.withDefaultTrace(state, mapErr, this);
+          }
+          case REPORT_WARNING -> {
+            errorsEncountered++;
+            if (errorsEncountered > MAX_MAP_WARNINGS) {
+              valueToAdd = nothing;
+            } else {
+              var wrappedInWarn =
+                  Warning.attach(ctx, nothing, err.getPayload(), null, appendWarningNode);
+              valueToAdd = wrappedInWarn;
+            }
+          }
+          case NO_WRAP -> {
+            return err;
+          }
+        }
+      }
+      target.add(valueToAdd, warnsLib);
+    }
+    var vector = target.asVector(true);
+    if (errorsEncountered >= MAX_MAP_WARNINGS) {
+      var additionalWarnsBuiltin = ctx.getBuiltins().getBuiltinType(AdditionalWarnings.class);
+      long additionalWarnsCnt = errorsEncountered - MAX_MAP_WARNINGS;
+      var additionalWarns = additionalWarnsBuiltin.newInstance(additionalWarnsCnt);
+      var vecWithAdditionalWarns =
+          Warning.attach(ctx, vector, additionalWarns, null, appendWarningNode);
+      return vecWithAdditionalWarns;
+    } else {
+      return vector;
+    }
+  }
+
+  private OnProblems processOnProblemsArg(Object onProblems, TypesLibrary typesLib) {
+    var ctx = EnsoContext.get(this);
+    var problemBehaviorBuiltin = ctx.getBuiltins().getBuiltinType(ProblemBehavior.class);
+    var noWrapBuiltin = ctx.getBuiltins().getBuiltinType(NoWrap.class);
+    var typeError =
```

Comment:
Will this get constructed on each call, or will it be optimized out if it's not used?

Reply:
`onProblems` parameter is cached in 89b5e59f4d7fe38d063e4aeb598a138fc7aae350. It helps with the performance.